cvmfs/cache_ram.cc

```diff
@@ -0,0 +1,250 @@
+  bool sweep_tail = false;
+  {
+    ReadLockGuard guard(rwlock_);
+    if (!IsValid(fd)) return -EBADFD;
+    file_descriptor = open_fds_[fd];
+    assert(pinned_entries_.GetBuffer(file_descriptor.handle, &buf));
+  }
+
+  WriteLockGuard guard(rwlock_);
+  assert(pinned_entries_.Unref(file_descriptor.handle));
+  if (pinned_entries_.GetRefcount(file_descriptor.handle) == 0) {
+    switch (buf.object_type) {
+    case cache::CacheManager::kTypeRegular:
+      assert(pinned_entries_.PopBuffer(file_descriptor.handle, &buf));
+      assert(!regular_entries_.Commit(file_descriptor.handle, buf));
+      break;
+    case cache::CacheManager::kTypeVolatile:
+      assert(pinned_entries_.PopBuffer(file_descriptor.handle, &buf));
+      assert(!volatile_entries_.Commit(file_descriptor.handle, buf));
+      break;
+    case cache::CacheManager::kTypePinned:
+    case cache::CacheManager::kTypeCatalog:
+      // just leave it in the pinned cache
+      break;
+    }
+    open_fds_[fd].handle = invalid_fd_;
+    sweep_tail = (static_cast<unsigned>(fd) == (open_fds_.size() - 1));
+  }
+
+  if (sweep_tail) {
+    unsigned last_good_idx = open_fds_.size() - 1;
+    while (open_fds_[last_good_idx].handle != invalid_fd_)
+      last_good_idx--;
+    open_fds_.resize(last_good_idx + 1);
+  }
+
+  return 0;
+}
+
+
+int64_t RamCacheManager::Pread(
+  int fd,
+  void *buf,
+  uint64_t size,
+  uint64_t offset)
+{
+  ReadLockGuard guard(rwlock_);
+  if (!IsValid(fd)) return -EBADFD;
+  return pinned_entries_.Read(open_fds_[fd].handle, buf, size, offset);
+}
+
+
+int RamCacheManager::Dup(int fd) {
+  WriteLockGuard guard(rwlock_);
+  ReadOnlyFd file_descriptor;
+  if (!IsValid(fd)) return -EBADFD;
+  file_descriptor = open_fds_[fd];
+  assert(pinned_entries_.Ref(file_descriptor.handle));
+  return AddFd(file_descriptor);
+}
+
+
+/**
+ * For a RAM cache, read-ahead is a no-op.
+ */
+int RamCacheManager::Readahead(int fd) {
+  ReadLockGuard guard(rwlock_);
+  if (!IsValid(fd))
+    return -EBADFD;
+  return 0;
+}
+
+
+int RamCacheManager::StartTxn(const shash::Any &id, uint64_t size, void *txn) {
+  Transaction *transaction = new (txn) Transaction();
+  transaction->id = id;
+  transaction->pos = 0;
+  transaction->expected_size = size;
+  //TODO realloc() on write for kSizeUnknown?
+  transaction->size = (size == kSizeUnknown) ? kPageSize : size;
+  if (transaction->size) {
+    transaction->buffer = scalloc(1, transaction->size);
+  }
+  return 0;
+}
+
+
+void RamCacheManager::CtrlTxn(
+  const string &description,
+  const ObjectType type,
+  const int flags,
+  void *txn)
+{
+  Transaction *transaction = reinterpret_cast<Transaction *>(txn);
+  transaction->description = description;
+  transaction->object_type = type;
+}
+
+
+int64_t RamCacheManager::Write(const void *buf, uint64_t size, void *txn) {
+  Transaction *transaction = reinterpret_cast<Transaction *>(txn);
+  uint64_t copy_size= min(size, transaction->size - transaction->pos);
```

Comment:
If we new the size upfront, it is an error to write more than the expected size.